File: gauth-core/gauth_core/utils/checksums.py
"""SHA-256 checksum computation using canonical JSON serialization."""

from __future__ import annotations

import hashlib
import json
from typing import Any


def canonical_json(obj: Any) -> str:
    if isinstance(obj, dict):
        items = sorted(
            (k, canonical_json(v)) for k, v in obj.items()
        )
        return "{" + ",".join(f"{json.dumps(k)}:{v}" for k, v in items) + "}"
    elif isinstance(obj, (list, tuple)):
        return "[" + ",".join(canonical_json(v) for v in obj) + "]"
    elif isinstance(obj, bool):
        return "true" if obj else "false"
    elif obj is None:
        return "null"
    elif isinstance(obj, (int, float)):
        return json.dumps(obj)
    elif isinstance(obj, str):
        return json.dumps(obj)
    else:
        return json.dumps(str(obj))


def _sha256_hex(data: str) -> str:
    return "sha256:" + hashlib.sha256(data.encode("utf-8")).hexdigest()


def compute_scope_checksum(scope: dict[str, Any]) -> str:
    return _sha256_hex(canonical_json(scope))


def compute_tool_permissions_hash(core_verbs: dict[str, Any]) -> str:
    return _sha256_hex(canonical_json(core_verbs))


def compute_platform_permissions_hash(platform_permissions: dict[str, Any]) -> str:
    return _sha256_hex(canonical_json(platform_permissions))

File: lib/db/src/schema/checksums.ts
import { createHash } from "crypto";

function canonicalJson(obj: unknown): string {
  if (obj === null || obj === undefined) {
    return "null";
  }
  if (typeof obj === "boolean") {
    return obj ? "true" : "false";
  }
  if (typeof obj === "number") {
    if (Number.isNaN(obj)) {
      throw new Error("NaN is not allowed in canonical JSON (RFC 8785)");
    }
    if (!Number.isFinite(obj)) {
      throw new Error("Infinity is not allowed in canonical JSON (RFC 8785)");
    }
    return JSON.stringify(obj);
  }
  if (typeof obj === "string") {
    return JSON.stringify(obj);
  }
  if (Array.isArray(obj)) {
    return "[" + obj.map(canonicalJson).join(",") + "]";
  }
  if (typeof obj === "object") {
    const entries = Object.entries(obj as Record<string, unknown>)
      .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
      .map(([k, v]) => `${JSON.stringify(k)}:${canonicalJson(v)}`);
    return "{" + entries.join(",") + "}";
  }
  return JSON.stringify(String(obj));
}

function sha256Hex(data: string): string {
  return "sha256:" + createHash("sha256").update(data, "utf-8").digest("hex");
}

export function computeScopeChecksum(
  scope: Record<string, unknown>,
): string {
  return sha256Hex(canonicalJson(scope));
}

export function computeToolPermissionsHash(
  coreVerbs: Record<string, unknown>,
): string {
  return sha256Hex(canonicalJson(coreVerbs));
}

export function computePlatformPermissionsHash(
  platformPermissions: Record<string, unknown>,
): string {
  return sha256Hex(canonicalJson(platformPermissions));
}

export { canonicalJson };
